File: 3D_Recon_Launcher_for_Real_Data/XT_IOMisc.py
""" ============================================================================
 * Copyright (c) 2013 K. Aditya Mohan (Purdue University)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright notice, this
 * list of conditions and the following disclaimer in the documentation and/or
 * other materials provided with the distribution.
 *
 * Neither the name of K. Aditya Mohan, Purdue
 * University, nor the names of its contributors may be used
 * to endorse or promote products derived from this software without specific
 * prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ """


#! /usr/bin/python

from PIL import Image
from os import system
import numpy as np
import h5py

def error_by_flag (flag, message):
	if (flag != 0):
		print 'ERROR_FLAG: ' + str(flag) 
		print message
		exit()

def create_folder (path2folder):
	print 'create_folder: Will create ' + path2folder 
	flag = system('mkdir ' + path2folder)
	if (flag != 0):
		print 'create_folder: Deleting ' + path2folder
		rm_flag = system('rm -r ' + path2folder)
		error_by_flag(rm_flag, 'ERROR: cannot remove ' + path2folder)

		print 'do_reconstruction: Creating ' + path2folder 
		flag = system('mkdir ' + path2folder)
		error_by_flag(flag, 'ERROR: cannot create ' + path2folder + '. Delete folder manually.')


def convert_um2HU (Object):
	density_scale = 0.001**4
	att_scale = 100**4

	AIR_MASS_ATT_COEFF = 0.496372335005353 * att_scale
	WATER_MASS_ATT_COEFF = 0.521225397034623 * att_scale

	WATER_DENSITY = 1.0 * density_scale
	AIR_DENSITY = 0.001205 * density_scale
	HOUNSFIELD_WATER_MAP = 1000
	HOUNSFIELD_AIR_MAP = 0
	slope = (HOUNSFIELD_WATER_MAP-HOUNSFIELD_AIR_MAP)/(WATER_MASS_ATT_COEFF*WATER_DENSITY-AIR_MASS_ATT_COEFF*AIR_DENSITY)
	c = -slope*(AIR_MASS_ATT_COEFF*AIR_DENSITY)
	Object = Object*slope + c
	return Object


def convert_HU2uint8 (Object, maxval, minval):
#	maxval = np.amax(Object)
#	minval = np.amin(Object)
	Object = (255.0/(maxval - minval))*(Object - minval);
	Object[Object <= 0] = 0
	Object[Object >= 255.0] = 255.0
	return Object.astype(np.uint8)

def write_array2tif (filename, array):
#	print 'Average value of image is ' + str(np.mean(array));
	img = Image.fromarray(array)
	img.save(filename)

def read_slice_of_tiff_file (filename, slice_num):
	img_file = Image.open(filename)
	Nx,Ny = img_file.size
	img_crop = img_file.crop((0, slice_num, Nx, slice_num + 1))
	img_data = img_crop.getdata()
	img = np.asarray(img_data, dtype=np.uint16)
	return img

def write_tiff_from_object_bin_file (proj, recon, files):
	Object = np.zeros((recon['N_z'], recon['N_xy'], recon['N_xy']), dtype = np.float64, order = 'C')
	for i in range(len(recon['r'])):
                path2results = files['Launch_Folder'] + 'run_' + 'sigs_' + str(recon['sigma_s'][i]) + '_sigt_' + str(recon['sigma_t'][i]) + '_r_' + str(recon['r'][i]) + '_K_' + str(proj['K']) + '_N_theta_' + str(proj['N_theta']) + '_N_p_' + str(proj['recon_N_p']) + '/'
		for j in range(int(recon['Rtime_num'][i])):
			for k in range(recon['node_num']):
				zpernode = recon['N_z']/recon['node_num']	
				Object[k*zpernode:(k+1)*zpernode,:,:] = np.fromfile(path2results + 'object_n' + str(k) + '_time_' + str(j) + '.bin', dtype = np.float64, count = -1).reshape((zpernode, recon['N_xy'], recon['N_xy']), order = 'C')
			for k in range(recon['N_z']):
				slice_object = convert_um2HU(Object[k,:,:])
		#		print 'write_tiff_from_object_bin_file: Average value of time slice ' + str(j) + ' in HU is ' + str(np.mean(slice_object))
				write_array2tif(path2results + 'object_z_' + str(k) + '_time_' + str(j) + '.tif', convert_HU2uint8(slice_object, recon['maxHU'], recon['minHU']))
			print 'write_tiff_from_object_bin_file: Average value of time slice ' + str(j) + ' in um^-1 is ' + str(np.mean(Object))

def write_object2HDF (proj, recon, files):
	Object = np.zeros((recon['N_z'], recon['N_xy'], recon['N_xy']), dtype = np.float64, order = 'C')
        for i in range(len(recon['r'])):
                path2results = files['Launch_Folder'] + 'run_' + 'smooth_' + str(recon['smoothness'][i]) + '_T_' + str(recon['ZingerT'][i]) + '_r_' + str(recon['r'][i]) + '_K_' + str(proj['K']) + '_N_theta_' + str(proj['N_theta']) + '_N_p_' + str(proj['recon_N_p']) + '/'
                file = h5py.File(path2results + 'object.hdf5', 'w');
#               dset = file.create_dataset('object', (recon['Rtime_num'][i], recon['N_z'], recon['N_xy'], recon['N_xy']), dtype=np.float32, chunks=True, compression='lzf');
                dset = file.create_dataset('object', (recon['Rtime_num'][i], recon['N_z'], recon['N_xy'], recon['N_xy']), dtype=np.float32, chunks=True);
                dset_off = file.create_dataset('proj_offset', (proj['recon_N_r'], proj['recon_N_t']), dtype=np.float32, chunks=True)
		for k in range(recon['node_num']):
                	proj_offset = np.fromfile(path2results + 'proj_offset_n' + str(k) + '.bin', dtype = np.float64, count = -1).reshape((proj['recon_N_r'], proj['recon_N_t']/recon['node_num']), order = 'C')
                	dset_off[:,k*proj['recon_N_t']/recon['node_num']:(k+1)*proj['recon_N_t']/recon['node_num']] = proj_offset.astype(np.float32)
                for j in range(int(recon['Rtime_num'][i])):
			zpernode = recon['N_z']/recon['node_num']	
			for k in range(recon['node_num']):
                        	Object[k*zpernode:(k+1)*zpernode,:,:] = np.fromfile(path2results + 'object_n' + str(k) + '_time_' + str(j) + '.bin', dtype = np.float64, count = -1).reshape((zpernode, recon['N_xy'], recon['N_xy']), order = 'C')

                        Object[Object < 0] = 0
                        dset[j,:,:,:] = Object.astype(np.float32)
                file.close

File: Source_Code_3D/XT_IOMisc.c
/* ============================================================================
 * Copyright (c) 2013 K. Aditya Mohan (Purdue University)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright notice, this
 * list of conditions and the following disclaimer in the documentation and/or
 * other materials provided with the distribution.
 *
 * Neither the name of K. Aditya Mohan, Purdue
 * University, nor the names of its contributors may be used
 * to endorse or promote products derived from this software without specific
 * prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */


#include <stdio.h>
#include "XT_Constants.h"
#include "tiff.h"
#include "allocate.h"
#include "XT_Structures.h"

/*Converts values from HU to um^-1*/
Real_t convert_HU2um (Real_t val)
{
       Real_t slope_HU=(HOUNSFIELD_WATER_MAP-HOUNSFIELD_AIR_MAP)/(WATER_MASS_ATT_COEFF*WATER_DENSITY-AIR_MASS_ATT_COEFF*AIR_DENSITY)/HFIELD_UNIT_CONV_CONST;
       Real_t c_HU=-slope_HU*(AIR_MASS_ATT_COEFF*AIR_DENSITY*HFIELD_UNIT_CONV_CONST);

       return((val-c_HU)/slope_HU);
}

/*Appends values in 'img' to a binary file with name filename. dimensions of img are specified by dim1, dim2, dim3 and dim4*/
void Append2Bin (char *filename, int dim1, int dim2, int dim3, int dim4, Real_t* img, FILE *debug_file_ptr)
{
  FILE *fp;
  char file[100];
  sprintf(file,"%s.bin", filename);
  fp=fopen (file,"ab");
  if (fp == NULL){fprintf(debug_file_ptr, "ERROR: Write2Bin: Cannot open file %s\n", file);exit(1);}
  fwrite(img,sizeof(Real_t),dim1*dim2*dim3*dim4,fp);	
  fclose(fp);
}

/*Writes values in 'img' to a binary file with name filename. dimensions of img are specified by dim1, dim2, dim3 and dim4*/
void Write2Bin (char *filename, int dim1, int dim2, int dim3, int dim4, Real_t* img, FILE *debug_file_ptr)
{
  FILE *fp;
  char file[100];
  sprintf(file,"%s.bin", filename);
  fp=fopen (file,"wb");
  if (fp == NULL){fprintf(debug_file_ptr, "ERROR: Write2Bin: Cannot open file %s\n", file);exit(1);}
  fwrite(img,sizeof(Real_t),dim1*dim2*dim3*dim4,fp);	
  fclose(fp);
}

/*Reads values in 'img' to a binary file with name filename. dimensions of img are specified by dim1, dim2, dim3 and dim4*/
void Read4mBin (char *filename, int dim1, int dim2, int dim3, int dim4, Real_t* img, FILE *debug_file_ptr)
{
  	char file[100];
  	sprintf(file,"%s.bin", filename);
	FILE* fp;
	int32_t size, result;

	size = dim1*dim2*dim3*dim4;
	fp = fopen (file, "rb" );
	if (fp==NULL) {fprintf(debug_file_ptr, "ERROR: error in reading file %s\n", file); exit (1);}	
	result = fread (img, sizeof(Real_t), size, fp);
  	if (result != size) 
	{fprintf(debug_file_ptr, "ERROR: Number of elements read does not match required, required = %d, read = %d\n", size, result);}
	fclose(fp);
}

/*Writes values in 'img' to tiff files. dimensions of img are specified by height and width.
hounsfield_flag if set, converts all values to HU and scales appropriately before writing to tiff file*/
void Write2Tiff(char* filename, int height, int width, Real_t** img, int hounsfield_flag, FILE *debug_file_ptr)
{
	struct TIFF_img out_img;
	int i,j;
	Real_t pixel;
	Real_t maxpix,minpix,avgpix,scale;
	FILE *fp;
	char file[100];
	Real_t slope,c;

	if(hounsfield_flag==1){
		slope=(HOUNSFIELD_WATER_MAP-HOUNSFIELD_AIR_MAP)/(WATER_MASS_ATT_COEFF*WATER_DENSITY-AIR_MASS_ATT_COEFF*AIR_DENSITY)/HFIELD_UNIT_CONV_CONST;
		c=-slope*(AIR_MASS_ATT_COEFF*AIR_DENSITY*HFIELD_UNIT_CONV_CONST);
		for ( i = 0; i < height; i++ )
	 	 for ( j = 0; j < width; j++ ) 
			img[i][j]=slope*img[i][j]+c;
#ifdef DEBUG_EN
		fprintf(debug_file_ptr, "Write2Tiff: file=%s, Hounsfeld slope = %f, c = %f\n",filename, slope,c);
#endif
	}	
	get_TIFF ( &out_img, height, width, 'g' );
	avgpix=0;
		
	maxpix=img[0][0];minpix=maxpix;
	 for ( i = 0; i < height; i++ )
 	 for ( j = 0; j < width; j++ ) {
		maxpix=(maxpix<img[i][j])?img[i][j]:maxpix;
		minpix=(minpix>img[i][j])?img[i][j]:minpix;
		avgpix+=img[i][j];	
	}
	avgpix/=(height*width);
	fprintf(debug_file_ptr, "Write2Tiff:file=%s,maxpix=%f,minpix=%f,avgpix=%f,height=%d,width=%d\n",filename,maxpix,minpix,avgpix,height,width);

	if(hounsfield_flag==1){
		maxpix=HOUNSFIELD_MAX;
		minpix=HOUNSFIELD_MIN;
	}
	scale=255/(maxpix-minpix);
	 for ( i = 0; i < height; i++ )
 	 for ( j = 0; j < width; j++ ) {
    		pixel = (int32_t)((img[i][j]-minpix)*scale);
    		if(pixel>255) {
      			out_img.mono[i][j] = 255;
    		}
    		else {
      			if(pixel<0) out_img.mono[i][j] = 0;
      			else out_img.mono[i][j] = pixel;
    		}
  	}
	sprintf(file,"%s.tif", filename);	
	if ( ( fp = fopen ( file, "wb" ) ) == NULL ) {
    		fprintf ( stderr, "cannot open file %s\n",filename);
	        exit ( 1 );
  	}

	/* write image */
	if ( write_TIFF ( fp, &out_img ) ) {
    		fprintf ( stderr, "error writing TIFF file %s\n", filename );
		exit ( 1 );
  	}

  /* close image file */
  fclose ( fp );
}

/*Writes to tiff files from a int32 array*/
void WriteInt32Tiff(char* filename, int height, int width, int32_t** imgin, int hounsfield_flag, FILE *debug_file_ptr)
{
	Real_t** img;
	int32_t i, j;

	img = (Real_t**)multialloc(sizeof(Real_t), 2, height, width);
	for (i = 0; i < height; i++)
	for (j = 0; j < width; j++)
		img[i][j] = (Real_t)imgin[i][j];
	Write2Tiff(filename, height, width, img, hounsfield_flag, debug_file_ptr);
	multifree(img,2);
}

/*Writes a multi-dimension array in 'img' to tiff files with dimension given in dim[4].
dim2loop_1 and dim2loop_2 specifies the dimension over which we loop and write the tiff files.
dim2write_1 and dim2write_2 specifies the dimensions which are written to tiff files.*/
void WriteMultiDimArray2Tiff (char *filename, int dim[4], int dim2loop_1, int dim2loop_2, int dim2write_1, int dim2write_2, Real_t* img, int hounsfield_flag, FILE* debug_file_ptr)
{
	char file[100];
	Real_t** img_temp;
	int i,j,temp,k,l;

	if (dim2loop_1 > dim2loop_2){
		temp = dim2loop_2;
		dim2loop_2 = dim2loop_1;
		dim2loop_1 = temp;
		printf("WARNING: WriteMultiDimArray2Tiff: Swapping variables dim2write_1 and dim2write_2\n");
	}
 		
	fprintf(debug_file_ptr, "WriteMultiDimArray2Tiff: Writing to tiff file %s*.tif \n", filename);
    img_temp = (Real_t**) multialloc(sizeof(Real_t), 2, dim[dim2write_1], dim[dim2write_2]);
	for (i=0; i<dim[dim2loop_1]; i++){
	for (j=0; j<dim[dim2loop_2]; j++){
	for (k=0; k<dim[dim2write_1]; k++){
	for (l=0; l<dim[dim2write_2]; l++){
		if (dim2write_1 == 2 && dim2write_2 == 3)
			img_temp[k][l] = img[((i*dim[1]+j)*dim[2]+k)*dim[3]+l];
		else if (dim2write_1 == 1 && dim2write_2 == 3)
			img_temp[k][l] = img[((i*dim[1]+k)*dim[2]+j)*dim[3]+l]; 	
		else if (dim2write_1 == 0 && dim2write_2 == 3)
			img_temp[k][l] = img[((k*dim[1]+i)*dim[2]+j)*dim[3]+l]; 
		else if (dim2write_1 == 1 && dim2write_2 == 2)
			img_temp[k][l] = img[((i*dim[1]+k)*dim[2]+l)*dim[3]+j]; 	
		else if (dim2write_1 == 0 && dim2write_2 == 2)
			img_temp[k][l] = img[((k*dim[1]+i)*dim[2]+l)*dim[3]+j]; 	
		else if (dim2write_1 == 0 && dim2write_2 == 1)
			img_temp[k][l] = img[((k*dim[1]+l)*dim[2]+i)*dim[3]+j]; 
		else
			printf("ERROR: WriteMultiDimArray2Tiff: Dimensions not recognized\n");
	}}
		sprintf(file, "%s_d1_%d_d2_%d",filename,i,j);
		Write2Tiff(file, dim[dim2write_1], dim[dim2write_2], img_temp, hounsfield_flag, debug_file_ptr);
	}
	}

	multifree (img_temp,2);
}

void write_ObjectProjOff2TiffBinPerIter (Sinogram* SinogramPtr, ScannedObject* ScannedObjectPtr, TomoInputs* TomoInputsPtr)
{
	int32_t i, dim[4];
	char object_file[100];
	char projOffset_file[100] = PROJ_OFFSET_FILENAME;

	sprintf(projOffset_file, "%s_n%d", projOffset_file, TomoInputsPtr->node_rank);
	for (i = 0; i < ScannedObjectPtr->N_time; i++)
	{
		sprintf (object_file, "%s_n%d_time_%d", OBJECT_FILENAME, TomoInputsPtr->node_rank, i);
		Write2Bin (object_file, 1, ScannedObjectPtr->N_z, ScannedObjectPtr->N_y, ScannedObjectPtr->N_x, &(ScannedObjectPtr->Object[i][1][0][0]), TomoInputsPtr->debug_file_ptr);
		dim[0] = 1; dim[1] = ScannedObjectPtr->N_z; dim[2] = ScannedObjectPtr->N_y; dim[3] = ScannedObjectPtr->N_x;
		if (TomoInputsPtr->Write2Tiff == 1)
			WriteMultiDimArray2Tiff (object_file, dim, 0, 1, 2, 3, &(ScannedObjectPtr->Object[i][1][0][0]), 0, TomoInputsPtr->debug_file_ptr);
			/*Changed above line so that output image is scaled from min to max*/
	}
	dim[0] = 1; dim[1] = 1; dim[2] = SinogramPtr->N_r; dim[3] = SinogramPtr->N_t;
	Write2Bin (projOffset_file, 1, 1, SinogramPtr->N_r, SinogramPtr->N_t, &(SinogramPtr->ProjOffset[0][0]), TomoInputsPtr->debug_file_ptr);
	if (TomoInputsPtr->Write2Tiff == 1)
		WriteMultiDimArray2Tiff (projOffset_file, dim, 0, 1, 2, 3, &(SinogramPtr->ProjOffset[0][0]), 0, TomoInputsPtr->debug_file_ptr);

}

/*Writes boolean array to tif files*/
void WriteBoolArray2Tiff (char *filename, int dim[4], int dim2loop_1, int dim2loop_2, int dim2write_1, int dim2write_2, bool* imgin, int hounsfield_flag, FILE* debug_file_ptr)
{
	Real_t* img;
	int32_t i;

	img = (Real_t*)get_spc(dim[0]*dim[1]*dim[2]*dim[3], sizeof(Real_t));
	for (i = 0; i < dim[0]*dim[1]*dim[2]*dim[3]; i++)
	{
		if (imgin[i] == true)
			img[i] = 1;
		else
			img[i] = 0;
	}

	WriteMultiDimArray2Tiff (filename, dim, dim2loop_1, dim2loop_2, dim2write_1, dim2write_2, img, hounsfield_flag, debug_file_ptr);
	free(img);
}

